### PR TITLE
Add basic features

### PR DESCRIPTION
**Pull request by kitsuyaazuma**
## Description

Added basic features of Broom.

## Motivation and Context

None so far.

## How has this been tested?

Manually build and deploy the image according to [the document.](https://book.kubebuilder.io/cronjob-tutorial/running)

```bash
make docker-build docker-push IMG=<some-registry>/<project-name>:tag
make deploy IMG=<some-registry>/<project-name>:tag
```

Then, create Broom with Secrets (for webhook URL) and CronJob that causes OOM.

```bash
kubectl apply -f config/samples/secret.yaml
kubectl apply -f config/samples/ai_v1alpha1_broom.yaml
kubectl apply -f config/samples/cronjob.yaml
```

Notifications will be sent to Slack.

## Screenshots (if appropriate):
スクリーンショット 2024-04-01 19 07 42

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: api/v1alpha1/broom_types.go [hunk 1]
```diff
@@ -17,19 +17,79 @@ limitations under the License.
...
+		m.Mul(int64(y))
+	}
+	return nil
+}
+
+type BroomRetryPolicy string
+
+const (
+	RetryAllowPolicy  BroomRetryPolicy = "Allow"
+	RetryForbidPolicy BroomRetryPolicy = "Forbid"
```
Comment (kitagry) [inline api/v1alpha1/broom_types.go]: Hmm, I think `retryPolicy: Allow, Forbid` is a bit confusing...

`restartPolicy: OnOOM, Never` like job.spec.restartPolicy. What do you think about this?

File: api/v1alpha1/broom_types.go [hunk 2]
```diff
@@ -17,19 +17,79 @@ limitations under the License.
...
 type BroomSpec struct {
 	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
 
-	// Foo is an example field of Broom. Edit broom_types.go to remove/update
-	Foo string `json:"foo,omitempty"`
+	Target      BroomTarget      `json:"target,omitempty"`
+	Adjustment  BroomAdjustment  `json:"adjustment"`
+	RetryPolicy BroomRetryPolicy `json:"retryPolicy"`
+	Webhook     BroomWebhook     `json:"webhook"`
```
Comment (kitagry) [inline api/v1alpha1/broom_types.go]: Is this is for slack? I think the name is a bit confusing. You should use `Slack`.

File: internal/controller/broom_controller.go [hunk 3]
```diff
@@ -47,16 +66,284 @@ type BroomReconciler struct {
...
+				if ref.Kind == "Job" {
+					jobOwnedOOMContainers[ref] = append(jobOwnedOOMContainers[ref], cs.Name)
+				}
+			}
+		}
+	}
+
+	if r.ResolvedJobs == nil {
+		r.ResolvedJobs = make(map[types.UID]struct{}, 0)
+	}
```
Comment (kitagry) [inline internal/controller/broom_controller.go]: I think this should be initialized in contructor or in SetupWithManager.

```diff
// internal/controller/broom_controller.go
+func New(mgr Manager) *BroomReconciler {
+	return &BroomReconciler{
+		Client: mgr.GetClient(),
+		Scheme: mgr.GetScheme(),
+		ResolvedJobs: make(map[types.UID]struct{}, 0)
+	}
+}

// cmd/main.go
-	if err = (&controller.BroomReconciler{
-		Client: mgr.GetClient(),
-		Scheme: mgr.GetScheme(),
-	}).SetupWithManager(mgr); err != nil {
+	if err = controller.New(mgr).SetupWithManager(mgr); err != nil {
```

File: internal/controller/broom_controller.go [hunk 4]
```diff
@@ -47,16 +66,284 @@ type BroomReconciler struct {
...
+	return nil
+}
+
+// retryJob creates Job for the failed Job with updated CronJob jobTemplate spec
+func (r *BroomReconciler) retryJob(ctx context.Context, cj *batchv1.CronJob, info cronJobOOMInfo) (string, error) {
+	log := log.FromContext(ctx)
+	randomString := random.GetRandomString(5)
+	retriedJobName := fmt.Sprintf("%s-retry-%s", info.LastFailedJob.Name, randomString)
+	annotations := map[string]string{
+		"rendezvous.m3.com/retried-by-broom": "true",
```
Comment (kitagry) [inline internal/controller/broom_controller.go]: you should change annotation

File: internal/controller/broom_controller.go [hunk 5]
```diff
@@ -47,16 +66,284 @@ type BroomReconciler struct {
...
+	}
+	if err := r.Create(ctx, &job); err != nil {
+		return "", fmt.Errorf("unable to create retried Job: %w", err)
+	}
+	log.Info("Retried Job", "name", job.Name)
+	return job.Name, nil
+}
+
+// isTargeted returns whether the Cronjob matches all the fields specified for the target or not
+func isTargeted(cj batchv1.CronJob, target aiv1alpha1.BroomTarget) bool {
```
Comment (kitagry) [inline internal/controller/broom_controller.go:267]: I want test.

**Review by kitagry — CHANGES_REQUESTED** [review]

File: internal/controller/broom_controller.go (hunk 3 above)
Comment (kitsuyaazuma) [reply internal/controller/broom_controller.go]: Definitely. Fixed in [18c95ff](https://github.com/m3dev/broom/pull/2/commits/18c95fffa4863a71adb238abf022db756012ea05).

File: api/v1alpha1/broom_types.go (hunk 1 above)
Comment (kitsuyaazuma) [reply api/v1alpha1/broom_types.go]: I agree. Using `OnOOM` and `Never` makes so much sense!
I'm a little worried that "restart" might be misunderstood as another launch "in the same spec", just like Pod's `job.spec.restartPolicy`. Any thoughts on this?

Comment (kitsuyaazuma) [reply api/v1alpha1/broom_types.go]: I'm starting to think "restart" is better after all. Fixed in [4f1e1ff](https://github.com/m3dev/broom/pull/2/commits/4f1e1ff3f246584a6492c01d187592936d6943cb).

File: api/v1alpha1/broom_types.go (hunk 2 above)
Comment (kitsuyaazuma) [reply api/v1alpha1/broom_types.go]: Fixed in [6ef1660](https://github.com/m3dev/broom/pull/2/commits/6ef1660dfdeaffaf0c62b6b269e3aeee36f2b271).

File: internal/controller/broom_controller.go (hunk 4 above)
Comment (kitsuyaazuma) [reply internal/controller/broom_controller.go]: Fixed in [027c7d2](https://github.com/m3dev/broom/pull/2/commits/027c7d2b6c4348c8afc6f94cc2d643215c4f9c70).

File: internal/controller/broom_controller.go (hunk 5 above)
Comment (kitsuyaazuma) [reply internal/controller/broom_controller.go:267]: Added a simple but useful test for the isTargeted() function in [6bfb7db](https://github.com/m3dev/broom/pull/2/commits/6bfb7dbaaf765b7a23aa0da4f084778cc5ab9686).

Tests of other logic will be added later

**Comment by kitsuyaazuma** [comment]
All fixes are covered and testing is complete!

スクリーンショット 2024-04-02 0 58 42

File: internal/controller/broom_controller_test.go [hunk 6]
```diff
@@ -79,6 +153,38 @@ var _ = Describe("Broom Controller", func() {
...
+				Name:      cronJobName,
+				Labels: map[string]string{
+					"m3.com/use-broom": "true",
+				},
+			}
+			res := isTargeted(*cronJob, target)
+			Expect(res).To(BeTrue(), "Expected CronJob to be targeted but it is not")
+
+			emptyTarget := aiv1alpha1.BroomTarget{}
+			res = isTargeted(*cronJob, emptyTarget)
```
Comment (kitagry) [inline internal/controller/broom_controller_test.go]: I think ginkgo is for E2E test. So, it is preferable to simple go test for isTargeted.

Comment (kitsuyaazuma) [reply internal/controller/broom_controller_test.go]: That's right. I will add tests in other PRs.

**Review by kitagry — APPROVED** [review]

**Pull request merged**